### ISSUE TITLE
Destroy of a record not indexed yet is causing an error

### ISSUE THREAD

**Issue by ausangshukla** [pat/thinking-sphinx#753]
Destroy of a record not indexed yet is causing an error

We index our users periodically so some users newly added have not yet been indexed. When such a non indexed user is destroyed TS fails and the transaction is rolled back

```
User.last.destroy

SQL (0.3ms)  DELETE FROM `users` WHERE `users`.`id` = 3736
  Sphinx  Retrying query "UPDATE user_core
SET sphinx_deleted = 1
WHERE id IN (59782)
" after error: index user_core: index not available
  Sphinx  Retrying query "UPDATE user_core
SET sphinx_deleted = 1
WHERE id IN (59782)
" after error: index user_core: index not available
   (46.1ms)  ROLLBACK

```

Note that when the index is rebuilt the User.last.destroy works as expected

**Comment by justincscott16** [comment]
Hello @pat! I know this issue is quite old, but I am having trouble destroying User records in my Rails production app because of this problem and was wondering if you could help me out! The stack trace when I try to delete a User is the following:

```bash
pry(main)> u.destroy
   (0.3ms)  BEGIN
  Workplace Load (0.4ms)  SELECT `workplaces`.* FROM `workplaces` WHERE `workplaces`.`user_id` = 232 LIMIT 1
  SQL (0.3ms)  DELETE FROM `workplaces` WHERE `workplaces`.`id` = 157
  Event Load (0.4ms)  SELECT `events`.* FROM `events` WHERE `events`.`user_id` = 232
  SQL (0.6ms)  DELETE FROM `events` WHERE `events`.`id` = 156596
  SQL (0.2ms)  DELETE FROM `events` WHERE `events`.`id` = 156597
  SQL (0.2ms)  DELETE FROM `events` WHERE `events`.`id` = 156598
  SQL (0.2ms)  DELETE FROM `events` WHERE `events`.`id` = 156599
  SQL (0.2ms)  DELETE FROM `events` WHERE `events`.`id` = 156600
  SQL (0.2ms)  DELETE FROM `events` WHERE `events`.`id` = 156601
  SQL (0.2ms)  DELETE FROM `events` WHERE `events`.`id` = 156602
  SQL (0.2ms)  DELETE FROM `events` WHERE `events`.`id` = 156603
  SQL (0.2ms)  DELETE FROM `events` WHERE `events`.`id` = 156604
  SQL (0.2ms)  DELETE FROM `events` WHERE `events`.`id` = 156605
  SQL (0.2ms)  DELETE FROM `events` WHERE `events`.`id` = 156607
  SQL (0.2ms)  DELETE FROM `events` WHERE `events`.`id` = 156608
  SQL (0.2ms)  DELETE FROM `events` WHERE `events`.`id` = 156609
  PostReview Load (0.3ms)  SELECT `post_reviews`.* FROM `post_reviews` WHERE `post_reviews`.`user_id` = 232
  Practitioner Load (0.3ms)  SELECT `practitioners`.* FROM `practitioners` WHERE `practitioners`.`user_id` = 232
  Event Load (0.3ms)  SELECT `events`.* FROM `events` WHERE `events`.`resource_id` = 148 AND `events`.`resource_type` = 'Practitioner'
  KeySpecialty Load (0.3ms)  SELECT `key_specialties`.* FROM `key_specialties` WHERE `key_specialties`.`practitioner_id` = 148
  SQL (1.0ms)  DELETE FROM `key_specialties` WHERE `key_specialties`.`id` = 248
  SQL (0.4ms)  DELETE FROM `key_specialties` WHERE `key_specialties`.`id` = 249
  SQL (0.2ms)  DELETE FROM `key_specialties` WHERE `key_specialties`.`id` = 250
  PreReview Load (0.3ms)  SELECT `pre_reviews`.* FROM `pre_reviews` WHERE `pre_reviews`.`practitioner_id` = 148
  PostReview Load (0.2ms)  SELECT `post_reviews`.* FROM `post_reviews` WHERE `post_reviews`.`practitioner_id` = 148
  Referral Load (0.2ms)  SELECT `referrals`.* FROM `referrals` WHERE `referrals`.`practitioner_id` = 148
  Treatment Load (0.2ms)  SELECT `treatments`.* FROM `treatments` WHERE `treatments`.`practitioner_id` = 148
  SQL (0.2ms)  DELETE FROM `treatments` WHERE `treatments`.`id` = 142
  SQL (0.3ms)  DELETE FROM `practitioners` WHERE `practitioners`.`id` = 148
  Sphinx  Retrying query "UPDATE practitioner_delta
SET sphinx_deleted = 1
WHERE id IN (889)
" after error: index practitioner_delta: index not available
  Sphinx  Retrying query "UPDATE practitioner_delta
SET sphinx_deleted = 1
WHERE id IN (889)
" after error: index practitioner_delta: index not available
   (0.6ms)  ROLLBACK
ThinkingSphinx::SphinxError: index practitioner_delta: index not available
from /var/www/findwellness-production/shared/bundle/ruby/2.1.0/gems/thinking-sphinx-3.1.0/lib/thinking_sphinx/connection.rb:90:in `rescue in query'
```

I am using the follow:

Thinking Sphinx 3.1.0
Rails 4.0.11
Sphinx 2.1.5

Thanks for any help you can give if possible!

**Comment by pat** [comment]
Hi Justin - is this error happening every time? If so, sounds like there might be issues with your `practitioner_delta` index - perhaps running `rake ts:index` will help get that working again?

**Comment by justincscott16** [comment]
@pat I went into the root of the rails app on my server and ran `bundle exec rake ts:index --trace` and received the following:

```bash
** Invoke ts:index (first_time)
** Invoke environment (first_time)
** Execute environment
** Execute ts:index
Generating configuration to /var/www/findwellness-production/releases/20160829172109/config/sphinx.conf
rake aborted!
Errno::EEXIST: File exists @ dir_s_mkdir - /var/www/findwellness-production/releases/20160829172109/tmp/sphinx
/opt/ruby-2.1.0/lib/ruby/2.1.0/fileutils.rb:250:in `mkdir'
/opt/ruby-2.1.0/lib/ruby/2.1.0/fileutils.rb:250:in `fu_mkdir'
/opt/ruby-2.1.0/lib/ruby/2.1.0/fileutils.rb:224:in `block (2 levels) in mkdir_p'
/opt/ruby-2.1.0/lib/ruby/2.1.0/fileutils.rb:222:in `reverse_each'
/opt/ruby-2.1.0/lib/ruby/2.1.0/fileutils.rb:222:in `block in mkdir_p'
/opt/ruby-2.1.0/lib/ruby/2.1.0/fileutils.rb:208:in `each'
/opt/ruby-2.1.0/lib/ruby/2.1.0/fileutils.rb:208:in `mkdir_p'
/var/www/findwellness-production/shared/bundle/ruby/2.1.0/gems/thinking-sphinx-3.1.0/lib/thinking_sphinx/configuration.rb:92:in `render_to_file'
/var/www/findwellness-production/shared/bundle/ruby/2.1.0/gems/thinking-sphinx-3.1.0/lib/thinking_sphinx/rake_interface.rb:13:in `configure'
/var/www/findwellness-production/shared/bundle/ruby/2.1.0/gems/thinking-sphinx-3.1.0/lib/thinking_sphinx/rake_interface.rb:24:in `index'
/var/www/findwellness-production/shared/bundle/ruby/2.1.0/gems/thinking-sphinx-3.1.0/lib/thinking_sphinx/tasks.rb:9:in `block (2 levels) in <top (required)>'
/var/www/findwellness-production/shared/bundle/ruby/2.1.0/gems/rake-11.2.2/lib/rake/task.rb:248:in `call'
/var/www/findwellness-production/shared/bundle/ruby/2.1.0/gems/rake-11.2.2/lib/rake/task.rb:248:in `block in execute'
/var/www/findwellness-production/shared/bundle/ruby/2.1.0/gems/rake-11.2.2/lib/rake/task.rb:243:in `each'
/var/www/findwellness-production/shared/bundle/ruby/2.1.0/gems/rake-11.2.2/lib/rake/task.rb:243:in `execute'
/var/www/findwellness-production/shared/bundle/ruby/2.1.0/gems/rake-11.2.2/lib/rake/task.rb:187:in `block in invoke_with_call_chain'
/opt/ruby-2.1.0/lib/ruby/2.1.0/monitor.rb:211:in `mon_synchronize'
/var/www/findwellness-production/shared/bundle/ruby/2.1.0/gems/rake-11.2.2/lib/rake/task.rb:180:in `invoke_with_call_chain'
/var/www/findwellness-production/shared/bundle/ruby/2.1.0/gems/rake-11.2.2/lib/rake/task.rb:173:in `invoke'
/var/www/findwellness-production/shared/bundle/ruby/2.1.0/gems/rake-11.2.2/lib/rake/application.rb:152:in `invoke_task'
/var/www/findwellness-production/shared/bundle/ruby/2.1.0/gems/rake-11.2.2/lib/rake/application.rb:108:in `block (2 levels) in top_level'
/var/www/findwellness-production/shared/bundle/ruby/2.1.0/gems/rake-11.2.2/lib/rake/application.rb:108:in `each'
/var/www/findwellness-production/shared/bundle/ruby/2.1.0/gems/rake-11.2.2/lib/rake/application.rb:108:in `block in top_level'
/var/www/findwellness-production/shared/bundle/ruby/2.1.0/gems/rake-11.2.2/lib/rake/application.rb:117:in `run_with_threads'
/var/www/findwellness-production/shared/bundle/ruby/2.1.0/gems/rake-11.2.2/lib/rake/application.rb:102:in `top_level'
/var/www/findwellness-production/shared/bundle/ruby/2.1.0/gems/rake-11.2.2/lib/rake/application.rb:80:in `block in run'
/var/www/findwellness-production/shared/bundle/ruby/2.1.0/gems/rake-11.2.2/lib/rake/application.rb:178:in `standard_exception_handling'
/var/www/findwellness-production/shared/bundle/ruby/2.1.0/gems/rake-11.2.2/lib/rake/application.rb:77:in `run'
/var/www/findwellness-production/shared/bundle/ruby/2.1.0/gems/rake-11.2.2/exe/rake:27:in `<top (required)>'
/var/www/findwellness-production/shared/bundle/ruby/2.1.0/bin/rake:23:in `load'
/var/www/findwellness-production/shared/bundle/ruby/2.1.0/bin/rake:23:in `<main>'
Tasks: TOP => ts:index
```

I also tried changing into the config directory of the Rails app (thinking it may need to be where the `sphinx.conf` file is) and receive the same.

**Comment by pat** [comment]
There are (at least) two potential issues here: do the Sphinx files all belong to the same operating system user that runs the Rails app? And that you're using to run the rake task?

Secondly: do you have TS configured to use shared paths for configuration and indices locations? (i.e. what's in your `config/thinking_sphinx.yml` file for this environment?) Using shared paths means Sphinx data is in a consistent location, rather than being different for every deploy.

**Comment by justincscott16** [comment]
The settings in the Production are as followed:

```yml
settings: &settings
  utf8: true
  configuration_file: "<%= Rails.root %>/config/sphinx.conf"
  indices_location:   "<%= Rails.root %>/db/indices"
  stopwords:          "<%= Rails.root %>/config/search/stopwords.txt"
  exceptions:         "<%= Rails.root %>/config/search/exceptions.txt"
  log:                "<%= Rails.root %>/log/searchd.log"
  query_log:          "<%= Rails.root %>/log/searchd.query.log"
  pid_file:           "<%= Rails.root %>/tmp/pids/searchd.pid"
  binlog_path:        "<%= Rails.root %>/tmp/sphinx/binlog"

development:
  <<: *settings

test:
  <<: *settings

staging:
  <<: *settings

production:
  <<: *settings
```

Also, if you don't mind me asking, what do you mean by 

> do the Sphinx files all belong to the same operating system user that runs the Rails app? And that you're using to run the rake task?

I use Capistrano to deploy to the server and then ssh into the server if that is helpful at all.

Thanks again for helping with an issue that is so old! I really appreciate it!

**Comment by pat** [comment]
Just to clarify that point: I presume all Capistrano operations related to deploys and rake tasks and such are done through a non-root user? Is it the same user in all of those situations? Is the searchd process being run by that user?

As for the contents of your configuration file - super useful to have those details, thanks :) I really don't recommend the configuration you've got, though!

* Using the same directories/port for both development and test means you can't have Sphinx running for both environments on your machine at the same time.
* Using `Rails.root` in your staging/production environments means that all Sphinx files will be different per deploy, as they'll be in `/var/www/findwellness-production/releases/TIMESTAMP`. It's _much_ better to use a shared directory instead, as documented in the deployment docs page: http://freelancing-gods.com/thinking-sphinx/deployment.html - so, using `/var/www/findwellness-production/shared/` instead of `Rails.root` would be wise.

Once upon a time, the docs suggested your release-focused approach, but that means that every deploy would also mean a semantic link in the file-system, and a restart of Sphinx. Using the shared directory means netter of those steps are required.

I'm not quite sure if that's part of the problem you're seeing here, but it could be related!

**Comment by justincscott16** [comment]
@pat Thank you, sorry for such a delayed response. I will look into reconfiguring the Sphinx and see if that will make it easier to use in my production and staging environments! Thanks a ton!

**Comment by justincscott16** [comment]
@pat Hello again! I am finally getting around this and have changed my config settings to use the shared directory! It seems to be working better, but I am now getting the following error:

`Permission denied, will not index. Try --rotate option.`

I know that you mentioned something about the root user earlier, but if I run `sudo bundle exec rake ts:index` it doesn't work either!

**Comment by pat** [comment]
Hi Justin - it sounds like it might be a file system permissions issue? All sphinx files should be owned by the same user you're running the Thinking Sphinx rake tasks with, and that are running other tasks/commands for hosting your app. So, unless _everything_ is running as root, you should never run a ts rake task as root/sudo.

If the files are all owned by the same user, and the tasks are being run by that same user, can you also confirm that the PID file is in a consistent location, and has the right value stored within it?

**Comment by justincscott16** [comment]
@pat yes so I believe I am on the correct user that I need to be, which is the 'deploy' user.

I went to check the PID file which is located in the shared directory, as you had mentioned above. The PID file contains:

`9530`

Here is also what I get returned when I look for the searchd processes:

```
deploy@findwellness:/var/www/findwellness-production/shared/tmp/pids$ ps aux | grep searchd
www-data  2836  0.0  0.0  29920    36 ?        S    Sep07   0:00 /usr/local/bin/searchd -c /var/www/findwellness-staging/current/config/sphinx.conf
www-data  2837  0.2  0.1 105012  5856 ?        Sl   Sep07 240:56 /usr/local/bin/searchd -c /var/www/findwellness-staging/current/config/sphinx.conf
root      9529  0.0  0.0  29920   568 ?        S    Sep07   0:00 searchd
root      9530  0.2  0.2 704624  9660 ?        Sl   Sep07 242:25 searchd
deploy   25132  0.0  0.0   9404  2152 pts/0    S+   17:31   0:00 grep --color=auto searchd
```

It looks like whats in the PID file is connected to root, where I am thinking it should be connected to the deploy process?

**Comment by justincscott16** [comment]
@pat I just realized after running `ps aux | grep searchd` again, the process listed under deploy was different, not sure if that means anything or not!

```
deploy@findwellness:/var/www/findwellness-production/shared/tmp/pids$ ps aux | grep searchd
www-data  2836  0.0  0.0  29920    36 ?        S    Sep07   0:00 /usr/local/bin/searchd -c /var/www/findwellness-staging/current/config/sphinx.conf
www-data  2837  0.2  0.1 105012  5856 ?        Sl   Sep07 240:58 /usr/local/bin/searchd -c /var/www/findwellness-staging/current/config/sphinx.conf
root      9529  0.0  0.0  29920   568 ?        S    Sep07   0:00 searchd
root      9530  0.2  0.2 704624 11272 ?        Sl   Sep07 242:26 searchd
deploy   25519  0.0  0.0   9404  2132 pts/0    S+   17:43   0:00 grep --color=auto searchd
```

**Comment by pat** [comment]
The fact that a searchd process is running under root suggests that might be the underlying problem. I'd start by killing all searchd processes, then running `rake ts:start` with the deploy user and see how things go from there.

Also, that last process in each list is the grepping that you're running each time, so that's expected to be different :)

**Comment by justincscott16** [comment]
@pat that helps a lot. I still believe there is something interesting with the user I have access to, but my problem was solved.

I killed all the searchd processes, and then went to run `rake ts:start` in the root of the Rails app and it did not work. So I went into the config file and ran the `searchd` command and it still did not work. After that I ran `sudo searchd` which started the process and in doing so ran all of the indexes. So I am now able to delete the user. I still have the permissions errors running the TS rake tasks, but it looks like killing the searchd processes and then starting it again with `sudo` reindexes and works out.

I don't know if that is what should be happening and may be a quick and a bit hacky of a fix but it allowed me to work with the data I needed to. Thanks again so much for all the help, you are awesome :+1:
